File: 0207-course-schedule/0207-course-schedule.cpp
class Solution {
public:
   bool checkCycle( int node ,vector<vector<int>> &graph, vector<int> & vis ,vector<int>&visDfs )
{
    vis[node] =1;
    visDfs[node] =1;

    for( int  i= 0 ;i< graph[node].size() ;i++ )
    {
        if( !vis[graph[node][i]])
        {
            checkCycle(graph[node][i], graph, vis, visDfs);
        }
        if(vis[graph[node][i]] && visDfs[graph[node][i]])
            return true;
    }
    visDfs[node]=0 ;
    return false;
}

bool isCycle (vector<vector<int>> &graph , int n)
{
    vector<int>vis(n ,0 );
    vector<int>visDfs(n,0);
    for( int i=0 ; i< n; i++ )
    {
        if( !vis[i])
        {
            if(checkCycle(i,graph,vis,visDfs))
                return true;
        }
    }
    return false;
}
    bool canFinish(int numCourses, vector<vector<int>>& prerequisites)
{
    vector<vector<int>> graph(numCourses);
    for( int i=0 ;i< prerequisites.size() ;i++)
    {
        graph[prerequisites[i][1]].push_back(prerequisites[i][0]);
    }
    if(isCycle(graph,numCourses))
        return false;
    return true;
}
};

File: 0207-course-schedule/0207-course-schedule.js
var isCyclic = function (adj, vis, finish, node) {
  vis[node] = 1;
  let cyclic = 0;
  for (let negh = 0; negh < adj[node].length; negh++) {
    let neighbor = adj[node][negh];
    if (vis[neighbor] === 0)
      cyclic = cyclic || isCyclic(adj, vis, finish, neighbor);
    else if (finish[neighbor] === 0) cyclic = 1;
  }
  finish[node] = 1;
  return cyclic;
};

var canFinish = function (numCourses, prerequisites) {
  let graph = new Array(numCourses);
  let visit = new Array(numCourses).fill(0);
  let finish = new Array(numCourses).fill(0);
  for (let i = 0; i < numCourses; i++) {
    graph[i] = [];
  }

  for (let i = 0; i < prerequisites.length; i++) {
    graph[prerequisites[i][1]].push(prerequisites[i][0]);
  }

  console.log(graph);

  for (let course = 0; course < numCourses; course++) {
    if (visit[course] === 0) {
      if (isCyclic(graph, visit, finish, course)) return false;
    }
  }
  return true;
};